autoload/LanguageClient.vim

```diff
@@ -548,6 +548,16 @@ function! LanguageClient#textDocument_rangeFormatting(...) abort
     return LanguageClient#Call('textDocument/rangeFormatting', l:params, l:callback)
 endfunction
 
+function! LanguageClient#completionItem_resolve(completion_item, ...)
```

Comment:
Where is this function get called?

Reply:
-> starts with `<CR>` key
-> `inoremap <silent> <expr> <CR> ncm2_snipmate#expand_or("\<CR>", 'n')`
-> [ncm2_snipmate#expand_or](https://github.com/ncm2/ncm2-snipmate/blob/e7b36231224eeae05355c7bc9f8c180426058cad/autoload/ncm2_snipmate.vim#L8)
-> [ncm2_snipmate#do_expand_or()](https://github.com/ncm2/ncm2-snipmate/blob/e7b36231224eeae05355c7bc9f8c180426058cad/autoload/ncm2_snipmate.vim#L17)
-> `<Plug>(ncm2_snipmate_expand_completed)`
-> [ncm2_snipmate#_do_expand_completed()](https://github.com/ncm2/ncm2-snipmate/blob/e7b36231224eeae05355c7bc9f8c180426058cad/autoload/ncm2_snipmate.vim#L45)
-> [apply_additional_text_edits](https://github.com/ncm2/ncm2-snipmate/blob/e7b36231224eeae05355c7bc9f8c180426058cad/pythonx/ncm2_lsp_snippet/utils.py#L2)
-> [call LanguageClient#completionItem_resolve](https://github.com/ncm2/ncm2-snipmate/blob/e7b36231224eeae05355c7bc9f8c180426058cad/pythonx/ncm2_lsp_snippet/utils.py#L13)
-> expand additional text edits
-> finally [call snipMate#expandSnip](https://github.com/ncm2/ncm2-snipmate/blob/e7b36231224eeae05355c7bc9f8c180426058cad/autoload/ncm2_snipmate.vim#L67) to expand snippets

According to [LSP specification](https://microsoft.github.io/language-server-protocol/specification). We need the whole `CompletionItem` passing to `completionItem/resolve`

```
Completion Item Resolve Request
The request is sent from the client to the server to resolve additional information for a given completion item.

Request:

method: ‘completionItem/resolve’
params: CompletionItem
Response:

result: CompletionItem
error: code and message set in case an exception happens during the completion resolve request.
```

Reply:
I see. I have 2 concerns,

1. Also from the protocol specification, the original CompletionItem might already have `additionalTextEdits`. Do you know which language server rely on a second `resolve` call to return `additionalTextEdits`?

2. If this is the way to go, it might be better just use CompletionItem as user_data here, as `textEdit`, `additionalTextEdits`, `snippet` etc are just part of it. You can leave this to me if you don't want to change too much.

Reply:
> Also from the protocol specification, the original CompletionItem might already have additionalTextEdits


IMO it would be meanless for a language server to have different `additionalTextEdits` in the first `CompletionItem` and the second `resolve`.

My current implementation:
- looks for `additionalTextEdits` in `CompletionItem`, 
- if empty, check `data` field in `CompletionItem`
- if has `data`, call `resolve`, and check `additionalTextEdits`
- apply additional text edits

> Do you know which language server rely on a second resolve call to return additionalTextEdits?

The vue language server needs a `resolve` to get additionalTextEdits.

> If this is the way to go, it might be better just use CompletionItem as user_data here, as textEdit, additionalTextEdits, snippet etc are just part of it

I think this would be more clean.  To be specific, we should send `CompletionItem` as `user_data.completion_item`,  there will be more plugins using user_data, keep it under a namespace is easier for completion plugins to detect where the item comes from.

After this PR getting merged, I'll change my plugins relying on `user_data.text_edit` and `user_data.snippet` to switch to `user_data.completion_item`.

Reply:
FYI, I asked for a demo in vue language server repo [here](https://github.com/vuejs/vetur/issues/839). It's easy  to setup vscode and test its behavior.

Reply:
FWIW, while not LSP, tsserver also has similar behavior where code actions are determined on a completionitem only after requesting additional completion details.